File: Weekly codes/2 SECOND WEEK CODES/ENVIRONMENT.jl
using Random
using StatsBase
using Images
using ImageFiltering
using Plots

# ==========================================
# --- GLOBAL PARAMETERS (1 px = 5 μm) ---
# ==========================================

# Total physical size remains 500μm x 500μm.
# Since 1 px = 5μm, the grid is now 100x100.
const GRID_SIZE = 100         
const CELL_DIAMETER_UM = 5    # 5μm diameter cells
const CELL_PIXEL_SIZE = 1     # 1 pixel represents exactly 5μm

const INITIAL_CELLS = 2     # Starting population
const INIT_PCD_PLUS_FRACTION = 0.50

const INIT_NUTRIENT_LEVEL = 100.0
const INIT_ANTIFUNGAL_LEVEL = 50.0

# ==========================================
# --- AGENT TYPES ---
# ==========================================

abstract type CellAgent end

mutable struct PCDPlus <: CellAgent
    x::Int
    y::Int
    alive::Bool
    biomass::Float64
end

mutable struct PCDMinus <: CellAgent
    x::Int
    y::Int
    alive::Bool
    biomass::Float64
end

# Constructors
PCDPlus(x, y; biomass=1.0) = PCDPlus(x, y, true, biomass)
PCDMinus(x, y; biomass=1.0) = PCDMinus(x, y, true, biomass)

# ==========================================
# --- MODEL STRUCTURE ---
# ==========================================

mutable struct PetriDishModel
    size::Int
    nutrient_layer::Matrix{Float64}
    ANTIFUNGAL_layer::Matrix{Float64}
    occupied_layer::Matrix{Bool} # Tracking pixel occupancy
    agents::Vector{CellAgent}
end

"""
    is_space_clear(model_size, occupied_layer, x, y)

Checks if the specific pixel (x, y) is available. 
In this scale (1px = 5um), a cell occupies exactly 1 pixel.
"""
function is_space_clear(model_size::Int, occupied_layer::Matrix{Bool}, x::Int, y::Int)
    if x < 1 || x > model_size || y < 1 || y > model_size
        return false # Out of bounds
    end
    # Since 1px = 1 cell, we just check if this specific pixel is taken
    return !occupied_layer[y, x]
end

function PetriDishModel()
    s = GRID_SIZE
    n_layer = fill(INIT_NUTRIENT_LEVEL, s, s)
    f_layer = fill(INIT_ANTIFUNGAL_LEVEL, s, s)
    occ_layer = zeros(Bool, s, s)
    
    agents = CellAgent[]
    num_plus = Int(round(INITIAL_CELLS * INIT_PCD_PLUS_FRACTION))

    # Initialize agents with pixel-level exclusion
    for i in 1:INITIAL_CELLS
        placed = false
        attempts = 0
        while !placed && attempts < 1000
            rx, ry = rand(1:s), rand(1:s)
            if is_space_clear(s, occ_layer, rx, ry)
                occ_layer[ry, rx] = true
                if i <= num_plus
                    push!(agents, PCDPlus(rx, ry))
                else
                    push!(agents, PCDMinus(rx, ry))
                end
                placed = true
            end
            attempts += 1
        end
    end

    return PetriDishModel(s, n_layer, f_layer, occ_layer, agents)
end

# ==========================================
# --- INITIAL STATE PLOTTING ---
# ==========================================

function plot_initial_state()
    println("Initializing 100x100 Environment (1px = 5μm)...")
    model = PetriDishModel()

    # Subplot 1: Nutrient Heatmap
    p1 = heatmap(model.nutrient_layer, 
                 title="Nutrients (Initial)", 
                 color=:viridis, 
                 aspect_ratio=:equal,
                 clims=(0, INIT_NUTRIENT_LEVEL))

    # Subplot 2: ANTIFUNGAL Heatmap
    p2 = heatmap(model.ANTIFUNGAL_layer, 
                 title="ANTIFUNGAL (Initial)", 
                 color=:ice, 
                 aspect_ratio=:equal,
                 clims=(0, INIT_ANTIFUNGAL_LEVEL))
    
    # Subplot 3: Cells (Pixel-based visualization)
    p3 = plot(title="Initial Cell Positions", 
              xlims=(0.5, model.size + 0.5), ylims=(0.5, model.size + 0.5), 
              aspect_ratio=:equal, legend=:outertopright)

    plus_x = [a.x for a in model.agents if isa(a, PCDPlus)]
    plus_y = [a.y for a in model.agents if isa(a, PCDPlus)]
    minus_x = [a.x for a in model.agents if isa(a, PCDMinus)]
    minus_y = [a.y for a in model.agents if isa(a, PCDMinus)]

    # Markersize is adjusted to look correct for a 100x100 grid
    if !isempty(plus_x)
        scatter!(p3, plus_x, plus_y, label="PCD+", color=:blue, markersize=4, markerstrokewidth=0.5)
    end
    if !isempty(minus_x)
        scatter!(p3, minus_x, minus_y, label="PCD-", color=:red, markersize=4, markerstrokewidth=0.5)
    end

    # Combine into a 3-pane layout
    final_plot = plot(p1, p2, p3, layout=(1, 3), size=(1400, 450))
    
    display(final_plot)
    println("Simulation grid of $(GRID_SIZE)x$(GRID_SIZE) plotted.")
end

plot_initial_state()

File: Weekly codes/2 SECOND WEEK CODES/ENVIRONMENT.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.ndimage import convolve
import random
from matplotlib.patches import Patch

#parameters
# Total physical size = 500μm x 500μm.
# 1 px = 5μm such that a cell takes up 1 px (100*100 grid)
GRID_SIZE = 100
TIME_STEP_DT = 1.0     # 1 hour per time step
SIMULATION_STEPS = 50

INITIAL_CELLS = 50
INIT_PCD_PLUS_FRACTION = 0.50

INIT_NUTRIENT_LEVEL = 100.0
INIT_ANTIFUNGAL_LEVEL = 50.0
DIFFUSION_NUTRIENT = 0.2       # D_n (Must be <= 0.25 for stability)
DIFFUSION_ANTIFUNGAL = 0.1    # D_f

# agents (the two genotpes are subclasses)
class CellAgent:
    def __init__(self, x, y, biomass=1.0):
        self.x = int(x)
        self.y = int(y)
        self.alive = True
        self.biomass = float(biomass)

class PCDPlus(CellAgent):
    pass

class PCDMinus(CellAgent):
    pass

# model
class PetriDishModel:
    def __init__(self):
        s = GRID_SIZE
        self.size = s
        self.nutrient_layer = np.full((s, s), INIT_NUTRIENT_LEVEL, dtype=float)
        self.antifungal_layer = np.full((s, s), INIT_ANTIFUNGAL_LEVEL, dtype=float)
        self.occupied_layer = np.zeros((s, s), dtype=bool)
        self.agents = []
        self.laplacian_kernel = np.array([[0,  1, 0],
                                          [1, -4, 1],
                                          [0,  1, 0]])

        num_plus = int(round(INITIAL_CELLS * INIT_PCD_PLUS_FRACTION))

        # placing cells randomly while ensuring no overlap (1 cell per pixel)
        for i in range(INITIAL_CELLS):
            placed = False
            attempts = 0
            while not placed and attempts < 10000:
                rx = np.random.randint(0, s)
                ry = np.random.randint(0, s)
                if self.is_space_clear(rx, ry):
                    self.occupied_layer[ry, rx] = True
                    if i < num_plus:
                        self.agents.append(PCDPlus(rx, ry))
                    else:
                        self.agents.append(PCDMinus(rx, ry))
                    placed = True
                attempts += 1

    def is_space_clear(self, x, y):
        if x < 0 or x >= self.size or y < 0 or y >= self.size:
            return False
        return not self.occupied_layer[y, x]
    
    def step_environment(self):
        laplacian_n = convolve(self.nutrient_layer, self.laplacian_kernel, mode='reflect')
        self.nutrient_layer += (DIFFUSION_NUTRIENT * laplacian_n) * TIME_STEP_DT
        self.nutrient_layer = np.clip(self.nutrient_layer, 0, None)

        laplacian_f = convolve(self.antifungal_layer, self.laplacian_kernel, mode='reflect')
        self.antifungal_layer += (DIFFUSION_ANTIFUNGAL * laplacian_f) * TIME_STEP_DT
        self.antifungal_layer = np.clip(self.antifungal_layer, 0, None)

# plotting 

def plot_initial_state():
    print("Initializing 100x100 Environment (1px = 5μm)...")
    model = PetriDishModel()

    fig, axes = plt.subplots(1, 3, figsize=(14, 4.5))

    # Subplot 1: Nutrient Heatmap
    im1 = axes[0].imshow(model.nutrient_layer, cmap='viridis', origin='lower', vmin=0, vmax=INIT_NUTRIENT_LEVEL, aspect='equal')
    axes[0].set_title('Nutrients)')
    fig.colorbar(im1, ax=axes[0], fraction=0.046, pad=0.04)

    # Subplot 2: ANTIFUNGAL Heatmap
    im2 = axes[1].imshow(model.antifungal_layer, cmap='Blues', origin='lower', vmin=0, vmax=INIT_ANTIFUNGAL_LEVEL, aspect='equal')
    axes[1].set_title('Antifungal agent)')
    fig.colorbar(im2, ax=axes[1], fraction=0.046, pad=0.04)

    # Subplot 3: Cells (Pixel-based visualization)
    axes[2].set_title('Petri dish')
    axes[2].set_xlim(-0.5, model.size - 0.5)
    axes[2].set_ylim(-0.5, model.size - 0.5)
    axes[2].set_aspect('equal')

    plus_x = [a.x for a in model.agents if isinstance(a, PCDPlus)]
    plus_y = [a.y for a in model.agents if isinstance(a, PCDPlus)]
    minus_x = [a.x for a in model.agents if isinstance(a, PCDMinus)]
    minus_y = [a.y for a in model.agents if isinstance(a, PCDMinus)]

    if plus_x:
        axes[2].scatter(plus_x, plus_y, label='PCD+', color='blue', s=16, linewidths=0.5)
    if minus_x:
        axes[2].scatter(minus_x, minus_y, label='PCD-', color='red', s=16, linewidths=0.5)

    axes[2].legend(loc='upper right', fontsize='small')

    plt.suptitle(f"Simulation grid of {GRID_SIZE}x{GRID_SIZE} (Initial)")
    plt.tight_layout()
    plt.show()

if __name__ == '__main__':
    plot_initial_state()
